sdk/core/core/inc/az_str.h

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef AZ_STR_H
+#define AZ_STR_H
+
+#include <az_assert.h>
+
+#include <az_static_assert.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <string.h>
+#include <limits.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// An immutable string slice.
+typedef struct {
+  // Points to the first byte (character).
+  uint8_t const *begin;
+  // A numer of octets (bytes) in the string.
+  size_t size;
+} az_const_str;
+
+// A muttable string slice.
+typedef struct {
+  uint8_t *begin;
+  size_t size;
+} az_str;
+
+AZ_STATIC_ASSERT(CHAR_BIT == 8);
+
+// A size of the string literal.
+// Details: to make sure that `S` is a `string literal`, we are appending `""` to `S`.
+#define AZ_STRING_LITERAL_LEN(S) (sizeof(S "") - 1)
+
+#define AZ_CONST_STR_DECL(NAME, STRING_LITERAL) \
+  static az_const_str const NAME = { .begin = STRING_LITERAL, .size = AZ_STRING_LITERAL_LEN(STRING_LITERAL) };
```

Comment:
Can the right hand side of the = be replaced with AZ_CONST_STR?
You'd have to reverse the two macros in this file.

Also, are we sure we really even want this macro? Maybe it's better for the developer to explicitly write
     static az_cost_str const NAME = AZ_CONST_STR("hi there")

so that they know exactly what is happening here. I worry that all these macros obfuscates the code. Someone who is unfamiliar with this macro will have no idea what it's doing. It's not always about reducing typing; it's about understanding what is happening so that when bugs happen, the code is understandable and fixable.

Reply:
The problem is C99. `AZ_CONST_STR` has `(az_const_str)` cast which makes the expression non-constant. For example, this code can't be compiled
```
static as_const_str hello = AZ_CONST_STR("hello");
```
VS error:
```
error C2099: initializer is not a constant
```

Reply:
But, if I understand this right, the code should not compile because we're trying to put a const value in a non-const variable. But, this would compile, right?

static as_const_str **_const_** hello = AZ_CONST_STR("hello");

Reply:
No, it wouldn't. The problem is not `const` but how C interprets the cast `(az_const_str)`. Here are more examples:
```c
typedef struct {
  int x;
} foo;

// Ok
foo const instance1 = { .x = 5 };

// Compilation error. error C2099: initializer is not a constant
foo const instance2 = (foo){ .x = 5 };

// Compiltion error. error C2099: initializer is not a constant
foor const instance3 = (foo const){ .x = 5 };

void func(foo const f);

// Ok
accept_foo((foo){ .x = 5 });

// Compilation error, syntax error: '{'
accept_foor({ .x = 5 });

foo create_foo();

// Ok
accpept_foo(create_foo());

// Compilation error. error C2099: initializer is not a constant
foo const instance4 = create_foo();
```